ext/drx/drx.c

```diff
@@ -126,10 +137,42 @@ drx_event_restore_state(void *drcontext, bool restore_memory,
                         dr_restore_state_info_t *info);
 #endif
 
+#ifdef PLATFORM_SUPPORTS_SCATTER_GATHER
+static per_thread_t *
+get_tls_data(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, tls_idx);
+    /* Support use during init (i#2910). */
+    if (pt == NULL)
+        return &init_pt;
+    return pt;
+}
+#endif
+
 /***************************************************************************
  * INIT
  */
 
+#ifdef PLATFORM_SUPPORTS_SCATTER_GATHER
+static void
+drx_thread_init(void *drcontext)
+{
+    per_thread_t *pt = (per_thread_t *)dr_thread_alloc(drcontext, sizeof(*pt));
+    pt->scratch_xmm_spill_slot = dr_thread_alloc(
+        dr_get_current_drcontext(), XMM_REG_SIZE + 31 /* for 32-byte alignment */);
```

Comment:
`drcontext` (else can have mismatches and problems)

Reply:
Add XMM_ALIGNMENT and use for all the 31 (-1) and 32 values

Reply:
Done.